### ISSUE TITLE
Cannot build geo as dependency

### ISSUE THREAD

**Issue by dmarteau** [georust/geo#647]
When using geo as dependency in another project the build fail with the following errors:

```
  Compiling geo v0.17.1 (/home/david/rust/geo/geo)
error[E0308]: mismatched types
   --> /home/david/rust/geo/geo/src/algorithm/line_intersection.rs:199:54
    |
199 |     let mut min_dist = point_line_euclidean_distance(p.start, q);
    |                                                      ^^^^^^^ expected struct `geo_types::Point`, found struct `geo_types::Coordinate`
    |
    = note: expected struct `geo_types::Point<_>`
               found struct `geo_types::Coordinate<F>`

error[E0308]: mismatched types
   --> /home/david/rust/geo/geo/src/algorithm/line_intersection.rs:201:46
    |
201 |     let dist = point_line_euclidean_distance(p.end, q);
    |                                              ^^^^^ expected struct `geo_types::Point`, found struct `geo_types::Coordinate`
    |
    = note: expected struct `geo_types::Point<_>`
               found struct `geo_types::Coordinate<F>`

error[E0308]: mismatched types
   --> /home/david/rust/geo/geo/src/algorithm/line_intersection.rs:206:46
    |
206 |     let dist = point_line_euclidean_distance(q.start, p);
    |                                              ^^^^^^^ expected struct `geo_types::Point`, found struct `geo_types::Coordinate`
    |
    = note: expected struct `geo_types::Point<_>`
               found struct `geo_types::Coordinate<F>`

error[E0308]: mismatched types
   --> /home/david/rust/geo/geo/src/algorithm/line_intersection.rs:211:46
    |
211 |     let dist = point_line_euclidean_distance(q.end, p);
    |                                              ^^^^^ expected struct `geo_types::Point`, found struct `geo_types::Coordinate`
    |
    = note: expected struct `geo_types::Point<_>`
               found struct `geo_types::Coordinate<F>`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `geo`
```

## How to reproduce:

* Create a new project `foobar` with `cargo init` 
* Include geo as dependency in Cargo.toml:
    ```
    [dependencies]
    geo  = { path = "../geo/geo" }
    ```
* Build  with `cargo build`

**Note** that building  geo directly from its repository run without errors.

I cannot figure what happends here, but I would suspect some scoped definitions not available when building outside the repository.

**Comment by michaelkirk** [comment]
This looks to be the common class of rust error where your project is unintentionally mixing-and-matching crate versions.

`cargo tree -d` can sometimes make this clearer.

In short, I suspect that because you are using a patched `geo` you'll also want to use a patched `geo-types`.

```
[dependencies]
geo  = { path = "../geo/geo" }
geo-types  = { path = "../geo/geo-types" }
```

Can you try that and let me know if you still have problems?

**Comment by dmarteau** [comment]
Updated the geo branch, and things magically solved !  Even if I revert to previous version, I have no issue anymore, go figure....

> In short, I suspect that because you are using a patched geo

No, I was using the genuine master branch.

I still did not figure out what happened.

**Comment by michaelkirk** [comment]
> Updated the geo branch, and things magically solved ! Even if I revert to previous version, I have no issue anymore, go figure....

This is probably because I have since released geo-types 0.7.2 and bumped geo's dependency on https://github.com/georust/geo/commit/bf07995eac6767dd45e186897798df8a98c2e38d

So master currently reflects a coherent release, but it won't always be so. Next time we do any dev on geo-types you'll have the same problem until we release it. 

>> In short, I suspect that because you are using a patched geo
>
> No, I was using the genuine master branch.

Sorry, I should have been clearer. I did not mean to imply that you had made changes to the geo-crate, what I meant is you'd added a `[patches.crates-io]` section to your cargo.toml to depend on an unreleased version of geo. That unreleased version of `geo` also required an unreleased version of `geo-types`, but you didn't tell your application to use the unreleased version of `geo-types`, hence the error.

Note that using unreleased versions of `geo` is not supported, so things might break for you, but if you insist on doing so, you'll have better luck if you also use the unreleased `geo-types`:

```
[dependencies]
geo  = { path = "../geo/geo" }
geo-types  = { path = "../geo/geo-types" }
```
